[CBRD-26251] Replication flag 가 설정된 클래스의 DML복제

youngjun9072 opened the pull request:
http://jira.cubrid.org/browse/CBRD-26251

### Purpose

클래스에 복제 플래그가 설정된(REPLICATION_OFF == false) 인 경우 DML을 슬레이브로 복제합니다.

### Implementation

- 복제 로그를 추가하기위한 조건으로 클래스에 복제 플래그가 설정되어있는지 확인합니다. 이를위해 heap_is_replication_class와 or_class_is_replication_on가 추가됩니다.
- 추가된 heap_is_replication_class는 데이터 입력(INSERT), 삭제(DELETEL)시 호출되는 locator_add_or_remove_index_internal와 수정(UPDATE)시 호출되는 locator_update_index의 조건식에 추가됩니다.
- 슬레이브에서는 데이터 반영을 준비하는 함수(locator_repl_prepare_force)에서 PK뿐만 아니라 UK 도 btree에서 찾을 수 있도록 수정합니다. 이와 동시에 함수의 이름도 btree_get_pkey_btid 에서 btree_get_rkey_btid로 변경됩니다.
- 복제 대상이되는 데이터는 PK 혹은 UK를 포함한 단일 컬럼 혹은 다중컬럼입니다.

테스트는 HA 환경에서 INSERT, UPDATE, DELTE를 수행하였고, 마스터, 슬레이브에서 값을 확인하였습니다. 
REPLICATION ON, RK 존재: 정상 복제
REPLICATION ON, RK 없음: 복제 제외
REPLICATION OFF, RK존재: 복제 제외
REPLICATION OFF, RK 없음: 복제 제외

### Remarks

복제 로그를 추가하기위한 조건식 중 need_replication, heap_is_replication_class(), !replicated는 서로 유사하면서도 다른 목적으로 사용되어 리팩토링이 필요해보입니다. 하지만 현 이슈에서는 replication 옵션에따른 DDL, DML 동작 여부가 우선이라고 생각되어 리팩토링 작업은 추후 진행하려합니다.

youngjun9072 committed:
add replication option

youngjun9072 committed:
lint

youngjun9072 committed:
fix typo

youngjun9072 committed:
Add variable intialization

youngjun9072 committed:
add "alter table ... replication" option

youngjun9072 committed:
lint

youngjun9072 committed:
add "TODO" comment

youngjun9072 committed:
add replication opt handling and select catalog

youngjun9072 committed:
Code changes

- Changed the location where replication option is handled
- Simplified condition logic to check only default or "on"
- Added explanatory comments

youngjun9072 committed:
lint

youngjun9072 committed:
replication 저장 및 조회 구현

youngjun9072 committed:
add alter table handling

youngjun9072 committed:
리뷰 반영
 - 매크로 이름 변경
 - 불필요 공백 삭제

youngjun9072 committed:
에러 핸들링 추가

youngjun9072 committed:
add tab

youngjun9072 committed:
add comment

youngjun9072 committed:
Flag and condition changes for backward compatibility with older backups

- replace SM_CLASSFLAG_REPLICATION with SM_CLASSFLAG_NO_REPLICATION.
- change handling condition from IS_REPLICATION_ON to !IS_REPLICATION.

youngjun9072 committed:
Add macro for alter condition branch

youngjun9072 committed:
Code modification

- Rename flag from SM_CLASSFLAG_REPLICATION to SM_CLASSFLAG_NO_REPLICATION
- Update return values to reflect the new flag naming
- Add error handling when attempting to alter replication tables in HA mode

youngjun9072 committed:
Add error message

youngjun9072 committed:
Merge branch 'feature/add-replication-option' into replication-option-handling

youngjun9072 committed:
remove unused error log

youngjun9072 committed:
Macro refactoring and lint

youngjun9072 committed:
Refactoring

youngjun9072 committed:
lint

youngjun9072 committed:
Apply review feedback
- fix indentation
- rename SM_CLASSFLAG_NO_REPLICATION to SM_CLASSFLAG_REPLICATION_OFF and update comment
- check replication status at function entry to avoid unnecessary operations
- adjust error message and parameter from 1 to 0
- add error handling

youngjun9072 committed:
replication pk only

youngjun9072 committed:
Implementation of unique key insert, delete, and update

youngjun9072 committed:
fix(pgbuf): prevent assertion in pgbuf_is_valid_page during UPDATE/DELETE

youngjun9072 committed:
lint

youngjun9072 commented:
/run sql medium

youngjun9072 committed:
Merge branch 'feature/add-replication-option' into replication_flag_only

youngjun9072 committed:
refectoring

youngjun9072 committed:
lint

hornetmj commented on src/base/object_representation_sr.c:
```diff
@@ -767,6 +767,30 @@ or_class_tde_algorithm (RECDES * record, TDE_ALGORITHM * tde_algo)
+  int flags = 0;
+  int replication_off_flag = 32;	/* SM_CLASSFLAG_NO_REPLICATION = 32 */
+
+  /* TODO:
```
코드 코멘트 보다는 EPIC 등에 기술하여 잊지 않고 검토되었으면 좋겠습니다.

youngjun9072 replied:
반영하였습니다. 추가적으로 TODO 코멘트에도 에픽 이슈 번호를 남겨 코드에서 TODO 발견 시 에픽도 함께 참조할수있도록하였습니다.

hornetmj commented on src/storage/btree.c:
```diff
@@ -8290,9 +8290,9 @@ btree_get_pkey_btid (THREAD_ENTRY * thread_p, OID * cls_oid, BTID * pkey_btid)
 	}
 
-      if (curr_idx->type == BTREE_PRIMARY_KEY)
+      if (curr_idx->type == BTREE_PRIMARY_KEY || curr_idx->type == BTREE_UNIQUE)
```
unique 인덱스 중 not null 제약이 있는 인덱스만 허용되어야 합니다. BTREE_UNIQUE를 이용해서 판단할 수 있을까요?

youngjun9072 replied:
not null 확인을 함께하는 함수 or_is_replication_key_candidate()를 추가하였습니다.

youngjun9072 committed:
Add replication key candidate checks and refactor related logic

- Add TODO with Epic reference
- Introduce `or_is_replication_key_candidate` function to check RK constraints
  (now requires NOT NULL condition for candidate attributes)
- Change return type of `heap_scancache_quick_start_root_hfid` to void
- Rename variable `pk_btid_index` to `rk_btid_index`
- Remove redundant `replicated` variable in update logic (replaced by `rk_btid_index`)
  - `rk_btid_index` is initialized to -1 and updated only once when a replication
    key candidate is found, then used after the loop for replication log handling

hgryoo commented on src/transaction/locator_sr.c:
```diff
@@ -8808,6 +8812,7 @@ locator_update_index (THREAD_ENTRY * thread_p, RECDES * new_recdes, RECDES * old
 	  error_code =
 	    repl_log_insert (thread_p, class_oid, oid, LOG_REPLICATION_DATA, RVREPL_DATA_UPDATE, repl_old_key,
 			     (REPL_INFO_TYPE) repl_info->repl_info_type);
+	  replicated = true;
```
이 부분 코드에 대해서 분석이 안되어서 단순히 근처 로직만 보고 질문 드립니다. replicated가 true로 설정되면, 다음 루프에서는 pk_btid_index가 이전 i 값으로 설정된 상태로 진행됩니다. 이 때, 위의 `if (pk_btid_index != -1)` 조건에 따라 동일한 btid index에 대해서 `repl_log_insert`에 또 도달할 가능성은 없을까요?

hgryoo replied:
상위의 for문에서 복제대상 BTID를 먼저 찾은 후 이 로직에 한번만 도달하는 것으로 확인.

H2SU commented on src/storage/heap_file.c:
```diff
@@ -11077,6 +11077,39 @@ heap_get_class_tde_algorithm (THREAD_ENTRY * thread_p, const OID * class_oid, TD
+      return false;
+    }
+
+  error = heap_scancache_quick_start_root_hfid (thread_p, &scan_cache);
```
반환되는 error가 여러 종류라면 return false가 아닌 return error; 또는 er_set으로 처리를 해야할 것 같습니다!

youngjun9072 replied:
NO_ERROR만 반환하는 함수이고, 다른 함수에서는 (void) heap_scancache_quick_start_root_hfid (thread_p, &scan_cache); 로 처리하고있어 동일하게 변경하였습니다.

hornetmj commented on msg/en_US.utf8/cubrid.msg:
```diff
@@ -1461,9 +1461,10 @@ Check the path of the key file (_keys) and make sure that it includes proper key
 1367 DBLINK Transaction ABORTED: %1$s
 
-1368 Changing the replication option of a class is not allowed in HA mode.
+1368 A replication key candidate (PRIMARY KEY or NOT NULL UNIQUE) is required in HA mode.
```
ko_KR.utf8에도 동일하게 추가해 주세요.

hornetmj replied:
L 다른 코드부 리뷰 진행하다가 메시지 내용의 적합성 리뷰를 빼먹었네요. 우선 다른 리뷰 내용 먼저 확인해 주세요.

hornetmj commented on src/base/object_representation_sr.c:
```diff
@@ -767,6 +767,30 @@ or_class_tde_algorithm (RECDES * record, TDE_ALGORITHM * tde_algo)
+or_class_is_replication_on (RECDES * record)
+{
+  int flags = 0;
+  int replication_off_flag = 32;	/* SM_CLASSFLAG_NO_REPLICATION = 32 */
```
```suggestion
  int replication_off_flag = 32;	/* SM_CLASSFLAG_REPLICATION_OFF = 32 */
```

youngjun9072 replied:
@hornetmj 반영했습니다.

hornetmj commented on src/base/object_representation_sr.c:
```diff
@@ -4690,3 +4714,39 @@ or_mvcc_get_prev_version_lsa (OR_BUF * buf, int mvcc_flags, LOG_LSA * prev_versi
+{
+  if (index == NULL)
+    {
+      return false;
```
```suggestion
      assert (false);
      return false;
```

youngjun9072 replied:
@hornetmj 반영했습니다.

hornetmj commented on src/base/object_representation_sr.c:
```diff
@@ -4690,3 +4714,39 @@ or_mvcc_get_prev_version_lsa (OR_BUF * buf, int mvcc_flags, LOG_LSA * prev_versi
+ *   index(in): OR_INDEX pointer (may be NULL)
+ */
+bool
+or_is_replication_key_candidate (const OR_INDEX * index)
```
후보키(candidate key)라는 표현이 더 많이 사용되고, 적합하다고 생각합니다.
```suggestion
or_is_replication_candidate_key (const OR_INDEX * index)
```

youngjun9072 replied:
@hornetmj 반영했습니다.

hornetmj commented on src/storage/btree.c:
```diff
@@ -8290,9 +8290,9 @@ btree_get_pkey_btid (THREAD_ENTRY * thread_p, OID * cls_oid, BTID * pkey_btid)
 	}
 
-      if (curr_idx->type == BTREE_PRIMARY_KEY)
+      if (or_is_replication_key_candidate (curr_idx))
```
Slave/Replica 노드의 index 순서에 의존하는 방식 대신 Master 노드에서 복제키명(replication key index name)을 전달해서 이 이름을 이용해서 적합한 인덱스를 찾는 접근이 더 적합하다고 생각합니다.

기본적인 HA 구성 및 검증 상에서는 제안하신 방법 역시 문제 없이 동작할 것으로 생각합니다. 다만, index 복제가 누락 또는 실패되는 경우나 특이한 구조로 HA를 구성하여 index 생성 순서와 복제 순서가 달라지게 되는 경우 등을 생각해 보면, index 순서와 조건(not null)에 의존하여 복제키를 찾기 보다는 마스터에서 전달 받은 복제키명을 사용하는게 더 안전합니다.

또한, 다음 확장성/안정성면 에서도 유리합니다.
1. 복제키를 사용자가 설정할 수 있는 경우 인덱스 생성 순서와 무관하게 됩니다.
2. not null unique 성질이 아닌 인덱스를 활용하게 되는 경우 (현재 가능성은 매우 낮음)
3. slave가 판단하지 않고, master가 남긴 정보를 따르는 접근이기 때문에 잘못된 인덱스를 선택하게 될 가능성 없어짐

hornetmj replied:
추가적으로, 최적화 관점에서 key(index name) - value(btid) 정보를 유지하는 것도 매번 탐색하는 것보다 이점이 있습니다. 본 이슈에서 고민하기 보다 TODO성으로 고민해 보면 좋을 것 같습니다.

youngjun9072 replied:
@hornetmj 복제키명을 사용하기위해서는 마스터의 서버,  applylogdb, 슬레이브의 서버까지 파악이되어야할것으로 생각됩니다. 제안주신 key(index name) - value(btid) 와 함께 나중 이슈에서 작업하는게 어떨까싶습니다.

hornetmj replied:
좋습니다. 해당 작업도 EPIC의 TODO에 기술해 주신 후 추후 별도 이슈로 처리 요망합니다.

youngjun9072 replied:
@hornetmj 주석으로 TODO와 에픽에 기술해놨습니다.

hornetmj commented on src/storage/heap_file.c:
```diff
@@ -11077,6 +11077,33 @@ heap_get_class_tde_algorithm (THREAD_ENTRY * thread_p, const OID * class_oid, TD
+      return false;
+    }
+
+  heap_scancache_end (thread_p, &scan_cache);
```
PEEK으로 읽었기 때문에 or_class_is_replication_on() -> heap_scancache_end() 순서로 호출되어야 합니다. (정보가 있는 page가 unfix 되기 전에 읽어야 합니다.)

youngjun9072 replied:
@hornetmj 반영했습니다.

hornetmj commented on src/transaction/locator_sr.c:
```diff
@@ -7785,6 +7785,7 @@ locator_add_or_remove_index_internal (THREAD_ENTRY * thread_p, RECDES * recdes,
   MVCCID mvccid;
   MVCC_REC_HEADER *p_mvcc_rec_header = NULL;
   bool classname_was_alloced = false;
+  bool replicated = false;
```
사용된 코드부를 보니 불필요해 보이는데, 어떤 의도가 있으신 건가요?

hornetmj commented on src/transaction/locator_sr.c:
```diff
@@ -8039,13 +8040,15 @@ locator_add_or_remove_index_internal (THREAD_ENTRY * thread_p, RECDES * recdes,
        */
-      if (need_replication && index->type == BTREE_PRIMARY_KEY && error_code == NO_ERROR
+      if (need_replication && heap_is_replication_class (thread_p, class_oid) && !replicated
+	  && or_is_replication_key_candidate (index) && error_code == NO_ERROR
```
or_is_replication_key_candidate() 함수로 검사 시 다음 조건이라면, 2번의 복제 로그가 생성됩니다. 코드는 이해를 돕기 위한 것으로 동작하는 sql문은 아닙니다. 이로 인해 insert 구문 하나가 2개의 복제 로그를 생성하고, 이는 두번의 복제 동작을 수행하며, 하나는 실패할 것입니다.
```
create table test (c1 int pk, c2 int unique key not null);
insert into test values (1, 1);
// 이 insert 구문은 pk와 uk 인덱스 모두를 갱신하며, 이에 따라 locator_add_or_remove_index_internal() 함수가 2회 호출됩니다.
```
적합한 접근은 이 시점에서는 복제키 후보가 아닌 이미 복제키로 정해진 인덱스인지 여부를 검사해야 합니다.

youngjun9072 replied:
@hornetmj  위에서 코맨트주신 replicated 플래그가 복제 로그의 반복 생성을 막기위한 용도입니다. 
인덱스들을 순회하며 복제 키가 선택되어 복제 로그를 생성하고나면 replicated 변수는 true로 바뀌기때문에 다음 루프부터는 복제키 후보가 존재해도 repl_log_insert가 호출되지 않도록하기위함입니다.

hornetmj replied:
이 부분도, 위 TODO 요청사항과 함께 정리하는게 좋을 것 같습니다(EPIC에 같이 기술해 주세요). RK의 모든 후보를 확인하고 replicated 플래그 처리하는 대신 이미 결정된 RK인 경우에만 복제 로그를 남기는 방식으로 처리 검토 요망합니다.

youngjun9072 replied:
@hornetmj 마찬가지로 주석으로 TODO와 에픽에 기술해놨습니다.

youngjun9072 committed:
Apply review feedback
- Corrected comment: SM_CLASSFLAG_NO_REPLICATION -> SM_CLASSFLAG_REPLICATION_OFF
- Changed the call order between or_class_is_replication_on() and heap_scancache_end()
- Added assert(false) for unexpected condition
- Rename function or_is_replication_key_candidate() -> or_is_replication_candidate_key()

hornetmj commented on msg/en_US.utf8/cubrid.msg:
```diff
@@ -1461,9 +1461,10 @@ Check the path of the key file (_keys) and make sure that it includes proper key
 
-1368 Changing the replication option of a class is not allowed in HA mode.
+1368 A replication key candidate (PRIMARY KEY or NOT NULL UNIQUE) is required in HA mode.
+1369 Changing the replication option of a class is not allowed in HA mode.
```
error_code.h 파일에 error code가 추가되지 않았습니다.

youngjun9072 replied:
@hornetmj error_code.h 에 클래스 옵션을 변경할 수 없는 에러인 ER_REPLICATION_CONSTRAINT 추가했습니다. 참고로 A replication key candidate... 에러 메시지는 제가 다른 작업을하는 과정에서 잘못 작성된 코드입니다.
추가로 메시지 한글 에러 메시지 추가했고 기존에 한글 에러메시지에서 코드가 잘못된 부분이있어 함께 수정했습니다.

hornetmj commented on msg/en_US.utf8/cubrid.msg:
```diff
@@ -1461,9 +1461,10 @@ Check the path of the key file (_keys) and make sure that it includes proper key
 
-1368 Changing the replication option of a class is not allowed in HA mode.
+1368 A replication key candidate (PRIMARY KEY or NOT NULL UNIQUE) is required in HA mode.
+1369 Changing the replication option of a class is not allowed in HA mode.
```
아래 코드에서 ER_REPLICATION_CONSTRAINT가 사용되고 있습니다.
1. er_set()에서 이미 msg 파일에 기술된 동일한 에러 메시지를 인자로 넘겨주고 있어서, 2중으로 동일한 메시지가 남습니다. er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, error, 0); 사용해 주시면 됩니다.
2. 1368번에 등록된 에러는 사용처가 없어 메시지 리뷰를 못했습니다.
```
10678 static int
10679 do_alter_change_replication (PARSER_CONTEXT * const parser, PT_NODE * const alter)
10680 {
10681   int error = NO_ERROR;
10682   const char *entity_name = NULL;
10683   DB_OBJECT *class_obj = NULL;
10684   DB_CTMPL *ctemplate = NULL;
10685   MOP class_mop = NULL;
10686   bool tran_saved = false;
10687   PT_NODE *replication_node = alter->info.alter.alter_clause.replication.tbl_replication;;
10688
10689   if (!HA_DISABLED () && IS_REPLICATION_ON_NODE (replication_node))
10690     {
10691       error = ER_REPLICATION_CONSTRAINT;
10692       er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, error, 0,
10693               "Changing the replication option of a class is not allowed in HA mode.");
```

youngjun9072 replied:
@hornetmj  
1. er_set() 에 에러 메시지 부분 제거했습니다.
2. 리뷰에서 언급하신 1368 는 A replication key candidate... 메시지로 다른 이슈 처리과정에서 잘못 작성된 코드라 삭제됐습니다. 현재는 1368이 ER_REPLICATION_CONSTRAINT에러입니다.

youngjun9072 committed:
Apply review feedback
- Removed redundant error specification in er_set
- Deleted unnecessary error messages
- Added Korean error message
- Added TODOs for additional tasks
